internal/clonepool/clonepool.go

```diff
@@ -0,0 +1,226 @@
+	if nbrClones+clonesToAdd > c.resource.ClonePool.MaxClones {
+		clonesToAdd = c.resource.ClonePool.MaxClones - nbrClones
+	}
+
+	for i := 0; i < clonesToAdd; i++ {
+		_, err := c.addCloneToPool(dss)
+		if err != nil {
+			fmt.Printf("error adding clone to pool %s\n", err.Error())
+			continue
+		}
+		// may be off a tiny bit of time
+		c.claimLock.Lock()
+		c.ClonesAvailable++
+		c.claimLock.Unlock()
+	}
+
+}
+
+func (c *ClonePool) getAvailableClones(dss *zfs.Dataset) ([]zdap.PublicClone, error) {
+	pooled, err := c.readPooled(dss)
+	if err != nil {
+		return nil, err
+	}
+
+	filtered := slicez.Filter(pooled, func(clone zdap.PublicClone) bool {
+		return clone.ExpiresAt == nil && clone.Healthy
+	})
+
+	return filtered, nil
+}
+
+func (c *ClonePool) addCloneToPool(dss *zfs.Dataset) (*zdap.PublicClone, error) {
+	snaps, err := c.cloneContext.GetResourceSnaps(dss, c.resource.Name)
+	if err != nil || snaps == nil {
+		return nil, err
+	}
+
+	sort.Slice(snaps, func(i, j int) bool {
+		return snaps[i].CreatedAt.Before(snaps[j].CreatedAt)
+	})
+	latestDate := snaps[0].CreatedAt
+
+	log.Infof("Adding clone to %s pool", c.resource.Name)
+	return c.cloneContext.CloneResourcePooled(dss, "zdapd", c.resource.Name, latestDate)
+}
+
+func (c *ClonePool) readPooled(dss *zfs.Dataset) ([]zdap.PublicClone, error) {
+	clones, err := c.cloneContext.Z.ListClones(dss)
+	if err != nil {
+		return nil, fmt.Errorf("could not list clones")
+	}
+	return slicez.Filter(clones, func(clone zdap.PublicClone) bool {
+		return clone.ClonePooled && clone.Resource == c.resource.Name
+	}), nil
+}
+
+func (c *ClonePool) pruneExpired(dss *zfs.Dataset, clones []zdap.PublicClone) []zdap.PublicClone {
+	c.claimLock.Lock()
+	defer c.claimLock.Unlock()
+
+	t := time.Now()
+	expired := slicez.Filter(clones, func(clone zdap.PublicClone) bool {
+		return clone.ExpiresAt != nil && clone.ExpiresAt.Before(t)
+	})
+
+	for _, e := range expired {
+		err := c.cloneContext.DestroyClone(dss, e.Name)
+		if err != nil {
+			fmt.Printf("could not destroy clone %s, error: %s", e.Name, err.Error())
+		}
+	}
+
+	return slicez.Filter(clones, func(clone zdap.PublicClone) bool {
+		return clone.ExpiresAt == nil || !clone.ExpiresAt.Before(t)
+	})
+}
+
+func (c *ClonePool) Expire(claimId string) error {
+	dss, err := c.cloneContext.Z.Open()
+	if err != nil {
+		return err
+	}
+
+	pooled, err := c.readPooled(dss)
+	if err != nil {
+		return err
+	}
+
+	match := slicez.Filter(pooled, func(a zdap.PublicClone) bool {
+		return a.Name == claimId
+	})
+
+	if len(match) == 0 {
+		return fmt.Errorf("found no matching clones")
+	}
+
+	c.cloneContext.Z.WriteLock()
+	err = match[0].Dataset.SetUserProperty(zfs.PropExpires, time.Now().Format(zfs.TimestampFormat))
+	c.cloneContext.Z.WriteUnlock()
+	return err
+}
+
+func (c *ClonePool) Claim(timeout time.Duration) (zdap.PublicClone, error) {
+	c.claimLock.Lock()
+	defer c.claimLock.Unlock()
+
+	dss, err := c.cloneContext.Z.Open()
```

Comment:
A bit scary to claim multiple locks like this. I think we're safe from deadlocks the way it works now though